twa_mops/molecular_fragment_utils.py

```diff
@@ -0,0 +1,1040 @@
+        vx = mx - c_pos.x
+        vy = my - c_pos.y
+        vz = mz - c_pos.z
+
+        dummy_x = mx + 1.0 * vx
+        dummy_y = my + 1.0 * vy
+        dummy_z = mz + 1.0 * vz
+
+        dummy_uid = str(uuid.uuid4())
+        data[dummy_uid] = {
+            "atom":         "X",
+            "coordinate_x": dummy_x,
+            "coordinate_y": dummy_y,
+            "coordinate_z": dummy_z,
+            "bond":         [],    # no explicit bonds
+            "mmtype":       "",
+            "qmmm":         "",
+        }
+
+def insert_pyrazole_binding_dummy_atoms(mol, conf, data):
+    """
+    Update CBU JSON dict to include dummy atoms for pyrozolate binding sites.
+    """
+    patt2 = Chem.MolFromSmarts('c1cn[n-]c1')
+    matches = mol.GetSubstructMatches(patt2)
+
+    for c1_idx, c2_idx, n1_idx, n2_idx, c3_idx in matches:
+        p1 = conf.GetAtomPosition(n1_idx)
+        p2 = conf.GetAtomPosition(n2_idx)
+        mx, my, mz = (p1.x + p2.x) / 2.0, (p1.y + p2.y) / 2.0, (p1.z + p2.z) / 2.0
+
+        # vector from carbon opposite nitrogens to midpoint of nitrogens
+        c_pos = conf.GetAtomPosition(c1_idx)
+        vx = mx - c_pos.x
+        vy = my - c_pos.y
+        vz = mz - c_pos.z
+
+        dummy_x = mx + 0.25 * vx
+        dummy_y = my + 0.25 * vy
+        dummy_z = mz + 0.25 * vz
+
+        dummy_uid = str(uuid.uuid4())
+        data[dummy_uid] = {
+            "atom":         "X",
+            "coordinate_x": dummy_x,
+            "coordinate_y": dummy_y,
+            "coordinate_z": dummy_z,
+            "bond":         [],    # no explicit bonds
+            "mmtype":       "",
+            "qmmm":         "",
+        }
+
+def cbu_mol_to_json_dict(mol: Chem.Mol, conf: Chem.Conformer) -> dict:
+    """
+    Convert an RDKit Mol and Conformer into a CBU JSON dict.
+    """
+
+    cbu_json = {}
+    idx2uid = {atom.GetIdx(): str(uuid.uuid4()) for atom in mol.GetAtoms()}
+
+    # atoms
+    for atom in mol.GetAtoms():
+        uid = idx2uid[atom.GetIdx()]
+        pos = conf.GetAtomPosition(atom.GetIdx())
+        cbu_json[uid] = {
+            "atom":         atom.GetSymbol(),
+            "coordinate_x": pos.x,
+            "coordinate_y": pos.y,
+            "coordinate_z": pos.z,
+            "bond":         [],
+            "mmtype":       atom.GetProp('_uffAtomType') if atom.HasProp('_uffAtomType') else "",
+            "qmmm":         "MM"
+        }
+
+    # bonds
+    for bond in mol.GetBonds():
+        a, b = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
+        order = 1.5 if bond.GetIsAromatic() else bond.GetBondTypeAsDouble()
+        cbu_json[idx2uid[a]]["bond"].append({"to_atom": idx2uid[b], "bond_order": order})
+        cbu_json[idx2uid[b]]["bond"].append({"to_atom": idx2uid[a], "bond_order": order})
+
+    # insert any binding site dummy atoms
+    insert_carboxylate_binding_dummy_atoms(mol, conf, cbu_json)
+    insert_pyrazole_binding_dummy_atoms(mol, conf, cbu_json)
```

Comment:
This seems super-arbitrary and specific, and also surprising, given that this function's intent appears to be generic. A comment explaining the reasoning/limitations behind this would be very helpful.